nested_forward_ad.py

```diff
@@ -0,0 +1,243 @@
+import torch
+from torch.autograd import forward_ad as fwAD
+from torch import Tensor
+from torch.utils._pytree import tree_map
+from torch.testing._internal.common_utils import (
+    TestCase, run_tests, disable_gc, parametrize,
+    instantiate_parametrized_tests
+)
+
+import functools
+import contextlib
+
+from utils import no_dispatch
+from base_tensor import BaseTensor
+
+# WARNING:
+# This class requires https://github.com/pytorch/pytorch/pull/73925 (that was reverted)
+# to properly work with forward AD implementation
+# If you get an error about "Expected this function to only be reached in inference mode"
+# then you don't have that patch!
+
+# This class wraps a pytorch dual Tensor and associates a level to it.
+# This allows to do multi-level forward AD even though pytorch only
+# support one level.
+class ForwardADTensor(BaseTensor):
```

Comment:
Is there a reason why we pass ForwardADTensor a PyTorch dual Tensor instead of a primal and a tangent?. It's a bit confusing now that there are "two types of dual tensors": we've got the PyTorch dual Tensor and we've got ForwardADTensor which is also a dual tensor (except it has a level associated with it)

Reply:
Yes, that's because the way we compute the Tangent is via the core forward AD. And that gives us a dual Tensor directly.
We could take the two separate here, but then that means that the torch_dispatch implementation would have to unpack dual and repack them here. So that here we would redo a make_dual with them.
That seems to be a lot of unpack/pack to get to exactly the same end state.